src/mito2/src/region/opener.rs

```diff
@@ -382,6 +399,30 @@ impl RegionOpener {
             return Ok(None);
         };
 
+        // Backfill partition_expr if missing.
+        {
+            let manifest = manifest_manager.manifest();
+            if manifest.metadata.partition_expr.is_none() {
+                let expr_json = self.partition_expr_fetcher.fetch_expr(self.region_id).await;
+                if let Some(expr_json) = expr_json {
+                    let metadata = manifest.metadata.as_ref().clone();
+                    let mut builder = RegionMetadataBuilder::from_existing(metadata);
+                    builder.partition_expr_json(Some(expr_json));
+                    let new_meta = Arc::new(builder.build().context(InvalidMetadataSnafu)?);
+
+                    let action = RegionMetaAction::Change(RegionChange { metadata: new_meta });
+                    // Use follower state during open.
+                    let _ = manifest_manager
+                        .update(
+                            RegionMetaActionList::with_action(action),
+                            RegionRoleState::Follower,
+                        )
+                        .await?;
```

Comment:
I wonder if this has the potential race condition in cluster mode.

Can we trigger a manifest checkpoint after the region becomes a leader intead?

Reply:
+1, The datanode may open the readonly regions